_posts/2018-06-18-On-Obsessive-Statelessness.markdown

```diff
@@ -0,0 +1,201 @@
+well.
+
+## Temporary Tokens
+
+The user/authentication JWT is not the only JWT in play in Peril.
+
+I needed the ability for the JavaScript Runtime to send messages back to the Peril server. Peril re-uses [JWTs for
+creating a short-lived (2 min) token][jwt_two]. This token only has access to a single installation and is given a
+list of mutations it has whitelisted access to in the GraphQL API. This token is a part of the data injected in at
+the start of the process. The Peril JavaScript DSL uses this token under the hood when you run particular functions.
+This the above post argues is a perfect use-case for JWTs.
+
+## Temporary Webhooks
+
+With an an admin user interface set up, you can now get a good overview of what your installation looks like in
+Peril.
+
+<img src="/images/peril-state/admin.png">
+
+This is a good start, but it's a static representation of a live system. In order to do any development of your
+dangerfiles in Peril, you would need to keep triggering the same event inside the GitHub and seeing how Peril
+evaluates your code. Event with the rich type definitions, you're unlikely to get it right first time.
+
+And here's where I made a compromise or two, in favour of a good abstraction. Ash recommended that perhaps storing
+webhooks from GitHub and making it feasible to re-send them in Peril would make a great development environment. I
+couldn't think of a way to do that statelessly, so I opted to time-box them. Now you can trigger a 5 minute window
+on an installation where any event sent to Peril will be stored in mongo for a week. After that they're gone.
+
+This is a great trade-off on data storage vs value of a feature. It's been the best idea so far on how to handle
+building a development mode into Peril, so I wouldn't want to compromise the feature in favour of something that
+won't store any data.
+
+## Real-time logging
+
+On the flip side, I spent a long time thinking about how I can get logs from a Peril run to a user without having to
+store those.
+
+I came up with what feels like such an obvious answer in retrospect. When you open up the admin dashboard, it
+connects to Peril via a websocket. This websocket is used to send real-time updates about when an event is
+triggering a dangerfile evaluation and its changing status. When the evaluation is finished, then the logs are
+collected and sent through the websocket to any users connected to the associated installation. The feature is
+particularly elegant because storing the logs for every Dangerfile run on something like S3 will not scale with my
+wallet. Plus, I don't want to have access to your logs ideally.
+
+<img src="/images/peril-state/websocket.png">
+
+Again, an interesting trade-off. You can only get logs when you're looking for them, as opposed to when a problem
+may actually have occurred. There's things I can do to work around this if and when it becomes a pressing need, for
+example keeping an installation's logs around for a day in Peril.
```

Comment:
Yeah totally, I wonder if you could integrate in to a user's Papertrail or something and move the problem entirely on to them? Like, provide a way for _them_ to record logs if they really want to instead of recording the logs yourself? Maybe!

Reply:
Hmm, yeah, if a user can use Peril to close tickets in Jira, if the logs are reified in some way (e.g. to a stream), maybe it's possible to provide a generic way to let the user do with them as they please.

Reply:
Agree on this, added